### PR TITLE
fix: Canvas jitters on page refresh on an application where widgets don't occupy the entire width of the canvas

### LINKED ISSUE #12292: [Bug]: Canvas jitters on page refresh on an application where widgets don't occupy the entire width of the canvas
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Description

Canvas jitters when widgets of differing widths are present and the page is refreshed. Seems to happen more strangely on apps where the widget width is not the entire length of the canvas

### Steps To Reproduce

1. Create a new application
2. Drop any widget and refresh the page. Observe the canvas jitter
3. Drop another widget of a different size and refresh the page. Observe the canvas jitter again
4. Drop more widgets or differing widths and refresh the page. Observe the canvas jitters even more with increasing widths of widgets

[LOOM DEMO](https://www.loom.com/share/581c1322417049bebc6a813449c7d2d9)

### Public Sample App

_No response_

### Version

Cloud

### PR DESCRIPTION
**Pull request by jsartisan**
When there are widgets on the canvas, on refresh the width of the widget flickers due to a bug in dynamic app layout logic. This PR fixes that

Fixes #12292

## Type of chang

- Bug fix

## How Has This Been Tested?

- Manually tested

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:green_circle: Total coverage has increased</summary>

    // Code coverage diff between base branch:release and head branch: fix/flickering-of-canvas-width 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | total | 56.64 **(0)** | 38.61 **(0.01)** | 35.96 **(0)** | 56.9 **(0.01)**
 :green_circle: | app/client/src/pages/Editor/WidgetsEditor/CanvasContainer.tsx | 100 **(2.38)** | 77.78 **(15.28)** | 100 **(0)** | 100 **(2.44)**
 :red_circle: | app/client/src/pages/Editor/WidgetsEditor/index.tsx | 93.55 **(-0.2)** | 60 **(0)** | 100 **(0)** | 93.22 **(-0.22)**
 :green_circle: | app/client/src/reducers/uiReducers/mainCanvasReducer.ts | 75 **(2.27)** | 22.22 **(0)** | 50 **(0)** | 75 **(2.27)**
 :sparkles: :new: | **app/client/src/selectors/mainCanvasSelectors.tsx** | **100** | **100** | **100** | **100**
 :green_circle: | app/client/src/utils/WorkerUtil.ts | 89.76 **(0)** | 72.55 **(1.96)** | 100 **(0)** | 93.33 **(0)**
 :red_circle: | app/client/src/utils/hooks/useDynamicAppLayout.tsx | 86.84 **(-1.47)** | 62.5 **(-3.21)** | 100 **(0)** | 90 **(0)**</details>

### CODE REVIEW

File: app/client/src/utils/hooks/useDynamicAppLayout.tsx [hunk 1]
```diff
@@ -181,13 +181,25 @@ export const useDynamicAppLayout = () => {
...
+      }
+    });
+
+    observer.observe(document, {
+      childList: true,
+      subtree: true,
+    });
   });
+
+  return initialized;
```
Comment (yaldram) [inline app/client/src/utils/hooks/useDynamicAppLayout.tsx:204]: we can return a clean up function `return () => { observer.disconnect() }`

File: app/client/src/utils/hooks/useDynamicAppLayout.tsx [hunk 2]
```diff
@@ -181,13 +181,25 @@ export const useDynamicAppLayout = () => {
...
+        return;
+      }
+    });
+
+    observer.observe(document, {
+      childList: true,
+      subtree: true,
+    });
   });
+
```
Comment (yaldram) [inline app/client/src/utils/hooks/useDynamicAppLayout.tsx:203]: Also we can pass an empty dependency array to useEffect `[]`

**Review by yaldram — CHANGES_REQUESTED** [review]

**Review by yaldram — APPROVED** [review]

**Review by keyurparalkar — APPROVED** [review]
LGTM!

**Comment by jsartisan** [comment]
@vivekverma2312 can someone from qa team pick this up to test?

**Comment by btsgh** [comment]
Tested on vercel link and found to be fixed.

**Pull request merged**